runtime/lua/vim/diagnostic.lua

```diff
@@ -901,22 +917,54 @@ M.handlers.signs = {
     end
 
     local ns = M.get_namespace(namespace)
-    if not ns.user_data.sign_group then
-      ns.user_data.sign_group = string.format('vim.diagnostic.%s', ns.name)
+    if not ns.user_data.sign_ns then
+      local _sign_ns = string.format('vim.diagnostic.%s', ns.name)
+      ns.user_data.sign_ns = vim.api.nvim_create_namespace(_sign_ns)
+    end
+
+    local sign_ns = ns.user_data.sign_ns
+
+    local function resolve_sign(diagnostic)
     end
 
-    local sign_group = ns.user_data.sign_group
     for _, diagnostic in ipairs(diagnostics) do
-      vim.fn.sign_place(0, sign_group, sign_highlight_map[diagnostic.severity], bufnr, {
-        priority = get_priority(diagnostic.severity),
-        lnum = diagnostic.lnum + 1,
-      })
+      local sign_text
+      if diagnostic.user_data.sign_text then
+        sign_text = diagnostic.user_data.sign_text
+      elseif opts.sign_text and opts.sign_text[diagnostic.severity] then
+        sign_text = opts.sign_text[diagnostic.severity]
+      else
+        sign_text = get_fallback_sign(diagnostic.severity) or M.severity[diagnostic.severity]:sub(1,1)
+      end
+
+      local sign_hl_group
+      if diagnostic.user_data.sign_hl_group then
+        sign_hl_group = diagnostic.user_data.sign_hl_group
+      else
+        sign_hl_group = sign_highlight_map[diagnostic.severity]
+      end
```

Comment:
You can use a single assignment style here.

```suggestion
      local sign_hl_group = diagnostic.user_data.sign_hl_group
                            or sign_highlight_map[diagnostic.severity]
```